File: ASCII_SUBSEQUENCES/Ascii_subsequences.java
import java.util.Scanner;

public class Ascii_Subsequences {
	
    public static void main(String args[]) {
	   
        /* In this problem we have to find all the possible Ascii subsequences
         * of the given string */
    
        Scanner obj = new Scanner(System.in);
	    String str = obj.next();
	    
	    int ans = find(str, ""); 
	
	    System.out.println();
	    System.out.println(ans); 
	
    }

    public static int find(String str, String ans) {
	    
        if (str.length() == 0) {
            
            System.out.print(ans + " ");
	        return 1;  		
	    }

        char ch = str.charAt(0);
        
        int x = find(str.substring(1), ans); // when ans is passed as it is
        int y = find(str.substring(1), ans + ch); // when ch is added to the ans
        int z = find(str.substring(1), ans + (int) ch); // when ch's ascii code is added to the ans

        return x + y + z; //the count for this frame will be sum of x,y,z
        
    }
}

// Sample input : ab
// Sample output :
//  b 98 a ab a98 97 97b 9798
// 9

File: Ascii_Subsequences/ASCII_Subsequences.js
/*ASCII SUBSEQUENCES PROBLEM
This problem is same as the "Subsequences" question just instead of two function calls, three function calls will be made :
- One with the same string.
- One with a letter added.
- One with the ascii code of the letter added to the string.
*/

function subseq(s, i, str) 
{
    if(i === str.length) 
    {
        console.log(s);
        return;
    }
    
    // The 3 function calls in the same order as explained above.
    // First function call
    subseq(s, i + 1, str); 
    // Second function call
    subseq(s + str[i], i + 1, str); 
    var n = str.charCodeAt(i); 
    // Third function call with the ASCII value of the ith element of the string
    subseq(s + n.toString(), i + 1, str); 
}

( function IIFE() {
    var str = prompt("Enter the string");
    console.log( "Number of Subsequences : ", Math.pow(3, str.length));
    subseq("", 0, str);
    return 0;
})();

/*Sample Input: Enter the string:
ab
Sample Output:
Number of Subsequences : 9

b
98
a
ab
a98
97
97b
9798
*/
